lib/msf/core/exploit/zeromq.rb

```diff
@@ -0,0 +1,122 @@
+
+  ZEROMQ_SIGNATURE = "\xff\x00\x00\x00\x00\x00\x00\x00\x01\x7f".freeze
+  ZEROMQ_VERSION = "\x03".freeze
+
+  include Msf::Exploit::Remote::Tcp
+
+  def zeromq_connect
+    print_status("Connecting to ZeroMQ service at #{peer}")
+    connect
+
+    # Replay the ZeroMQ elbow bump
+    zeromq_negotiate_signature
+    zeromq_negotiate_version
+    zeromq_negotiate_security_mechanism
+    zeromq_ready_req_router
+  end
+
+  def zeromq_negotiate_signature
+    print_status('Negotiating signature')
+
+    unless (signature = sock.get_once)
+      fail_with(Failure::Unknown, 'Did not receive signature')
+    end
+
+    unless signature == ZEROMQ_SIGNATURE
+      fail_with(Failure::UnexpectedReply,
+                "Received invalid signature: #{signature.inspect}")
+    end
+
+    vprint_good("Received valid signature: #{signature.inspect}")
+
+    vprint_status('Sending identical signature')
+    sock.put(signature)
+  end
+
+  def zeromq_negotiate_version
+    print_status('Negotiating version')
+
+    unless (version = sock.get_once)
+      fail_with(Failure::Unknown, 'Did not receive version')
+    end
+
+    unless version == ZEROMQ_VERSION
+      fail_with(Failure::UnexpectedReply,
+                "Received incompatible version: #{version.inspect}")
+    end
+
+    vprint_good("Received compatible version: #{version.inspect}")
+
+    vprint_status('Sending identical version')
+    sock.put(version)
+  end
+
+  def zeromq_negotiate_security_mechanism
+    print_status('Negotiating security mechanism')
+
+    unless (security = sock.get_once)
+      fail_with(Failure::Unknown, 'Did not receive security mechanism')
+    end
+
+    unless security.include?('NULL')
+      fail_with(Failure::UnexpectedReply,
+                "Did not receive NULL security mechanism: #{security.inspect}")
+    end
+
+    vprint_good('Received NULL security mechanism')
+
+    vprint_status('Sending identical security mechanism')
+    sock.put(security)
+  end
+
+  def zeromq_ready_req_router
+    print_status('Sending READY command of type REQ')
+
+    sock.put(
+      "\x04&\x05READY\vSocket-Type\x00\x00\x00\x03REQ\bIdentity\x00\x00\x00\x00"
+    )
+
+    unless (ready = sock.get_once)
+      fail_with(Failure::Unknown, 'Did not receive READY reply')
+    end
+
+    unless ready.match(/READY.+Socket-Type.+ROUTER.+Identity/)
+      fail_with(Failure::UnexpectedReply,
+                "Did not receive READY reply of type ROUTER: #{ready.inspect}")
+    end
+
+    vprint_good('Received READY reply of type ROUTER')
+  end
+
+  def zeromq_send_message(msg)
+    sock.put("\x01\x00\x02#{[msg.length].pack('Q>')}#{msg}")
+  end
+
+  # NOTE: This is Salt-specific code to serialize a cleartext MessagePack "load"
+  def serialize_clear_load(hash)
+    clear_load = {
+      'enc' => 'clear',
+      'load' => hash
+    }
+
+    # XXX: Strings NEED to be UTF-8 here, NOT binary!
+    # HACK: Bypass RuboCop with this one neat metaprogramming trick!
+    send(
```

Comment:
This double hack seems pretty gnarly; You can disable cops with:
https://docs.rubocop.org/en/stable/configuration/#disabling-cops-within-source-code

I'm guessing we're using eval to convert the hash into utf-8? Could just recursively convert the hash? Tried searching the codebase for an existing helper for this and didn't find it weirdly

Reply:
Might not be necessary now that I moved the code to a new file with the magic comment. UTF-8 should be fine in this file.

Reply:
Can't say I didn't enjoy the hack. :-)